pkg/query/exclusion.go

```diff
@@ -0,0 +1,102 @@
+package query
+
+import (
+	"github.com/authzed/spicedb/pkg/spiceerrors"
+)
+
+// Exclusion represents the set of relations that are in the mainSet but not in the excluded set.
+// This is equivalent to `permission foo = bar - baz`
+type Exclusion struct {
+	mainSet  Iterator
+	excluded Iterator
+}
+
+var _ Iterator = &Exclusion{}
+
+func NewExclusion(mainSet, excluded Iterator) *Exclusion {
+	return &Exclusion{
+		mainSet:  mainSet,
+		excluded: excluded,
+	}
+}
+
+func (e *Exclusion) CheckImpl(ctx *Context, resourceIDs []string, subjectID string) (RelationSeq, error) {
+	// Get all relations from the main set
+	mainSeq, err := ctx.Check(e.mainSet, resourceIDs, subjectID)
+	if err != nil {
+		return nil, err
+	}
+
+	mainRels, err := CollectAll(mainSeq)
+	if err != nil {
+		return nil, err
+	}
+
+	// If main set is empty, return empty result
+	if len(mainRels) == 0 {
+		return func(yield func(Relation, error) bool) {
+			// Empty sequence - never yield anything
+		}, nil
+	}
+
+	// Get all relations from the excluded set
+	excludedSeq, err := ctx.Check(e.excluded, resourceIDs, subjectID)
+	if err != nil {
+		return nil, err
+	}
+
+	excludedRels, err := CollectAll(excludedSeq)
+	if err != nil {
+		return nil, err
+	}
+
+	// Filter main set by excluding relations that are in the excluded set
+	return func(yield func(Relation, error) bool) {
+		for _, mainRel := range mainRels {
+			found := false
+
+			// Check if this relation exists in the excluded set (only compare endpoints: resource and subject object types/IDs)
+			for _, excludedRel := range excludedRels {
+				if mainRel.Resource.ObjectType == excludedRel.Resource.ObjectType &&
+					mainRel.Resource.ObjectID == excludedRel.Resource.ObjectID &&
+					mainRel.Subject.ObjectType == excludedRel.Subject.ObjectType &&
+					mainRel.Subject.ObjectID == excludedRel.Subject.ObjectID {
```

Comment:
This seems like it might be worth making into a `rel.Equal` method or something

Reply:
This is coming in the very next PR because this gets refactored